controllers/space/space_controller_test.go

```diff
@@ -0,0 +1,796 @@
+						}
+						return member1Client.Client.Delete(ctx, obj, opts...)
+					}
+					member1 := NewMemberClusterWithClient(member1Client, "member-1", corev1.ConditionTrue)
+					member2 := NewMemberCluster(t, "member-2", corev1.ConditionTrue)
+					ctrl := newReconciler(hostClient, member1, member2)
+
+					// when
+					res, err := ctrl.Reconcile(context.TODO(), reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Namespace: space.Namespace,
+							Name:      space.Name,
+						},
+					})
+
+					// then
+					require.NoError(t, err)
+					assert.Equal(t, reconcile.Result{Requeue: false}, res) // no need to explicitly requeue while the NSTemplate is terminating
+					nsTmplSet := nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, "oddity", member1.Client).
+						Exists().
+						HasDeletionTimestamp().
+						HasConditions(nstemplatetsettest.Terminating()).
+						Get()
+					spacetest.AssertThatSpace(t, test.HostOperatorNs, "oddity", hostClient).
+						Exists().
+						HasStatusTargetCluster("member-1").
+						HasConditions(spacetest.Terminating()).
+						HasFinalizer(toolchainv1alpha1.FinalizerName) // finalizer is still present while the NSTemplateSet is not fully deleted
+
+					t.Run("requeue while NSTemplateSet is terminating", func(t *testing.T) {
+						// given another round of requeue without while NSTemplateSet is *not ready*
+						nsTmplSet.Status.Conditions = []toolchainv1alpha1.Condition{
+							{
+								Type:   toolchainv1alpha1.ConditionReady,
+								Status: corev1.ConditionFalse,
+								Reason: toolchainv1alpha1.NSTemplateSetTerminatingReason,
+							},
+						}
+						err := member1.Client.Update(context.TODO(), nsTmplSet)
+						require.NoError(t, err)
+						ctrl := newReconciler(hostClient, member1, member2)
+
+						// when
+						res, err := ctrl.Reconcile(context.TODO(), reconcile.Request{
+							NamespacedName: types.NamespacedName{
+								Namespace: space.Namespace,
+								Name:      space.Name,
+							},
+						})
+
+						// then
+						require.NoError(t, err)
+						assert.Equal(t, reconcile.Result{Requeue: false}, res) // no need to explicitly requeue while the NSTemplate is terminating
+						// no changes
+						nsTmplSet := nstemplatetsettest.AssertThatNSTemplateSet(t, test.MemberOperatorNs, "oddity", member1.Client).
+							Exists().
+							HasDeletionTimestamp().
+							HasConditions(nstemplatetsettest.Terminating()).
+							Get()
+						spacetest.AssertThatSpace(t, test.HostOperatorNs, "oddity", hostClient).
+							Exists().
+							HasStatusTargetCluster("member-1").
+							HasConditions(spacetest.Terminating()).
+							HasFinalizer(toolchainv1alpha1.FinalizerName)
+
+						t.Run("done when NSTemplateSet is deleted", func(t *testing.T) {
+							// given another round of requeue without with NSTemplateSet now *fully deleted*
+							member1Client.MockDelete = nil
+							err := member1.Client.Delete(context.TODO(), nsTmplSet)
```

Comment:
by doing this, you more or less replicate the K8s logic. Wouldn't be better to configure the test properly and test the particular reconciliations in an isolated way?

Reply:
it's isolated per the use of chained subtest. I thought it would make sense to replicate the k8s logic in the unit tests, since it's the expected behaviour after all